dev/AppNotifications/AppNotificationUtility.cpp

```diff
@@ -232,79 +234,124 @@ std::wstring Microsoft::Windows::AppNotifications::Helpers::SetDisplayNameBasedO
+            localIconFilePath = wil::make_unique_string<wil::unique_cotaskmem_string>(localIconFilePathAsWstring.erase(iteratorForCommaDelimiter).c_str());
         }
 
-        auto iteratorForFileExtension{ iconFilePath.find_first_of(L".") };
+        *displayName = localDisplayName.release();
+        *iconFilePath = localIconFilePath.release();
+
+        return S_OK;
+    }
+
+    return HRESULT_FROM_WIN32(ERROR_NOT_FOUND);
+}
+
+void Microsoft::Windows::AppNotifications::Helpers::RegisterAssets(std::wstring const& appId, std::wstring const& clsid)
+{
+    wil::unique_hkey hKey;
+    // subKey: \Software\Classes\AppUserModelId\{AppGUID}
+    std::wstring subKey{ c_appIdentifierPath + appId };
+
+    THROW_IF_WIN32_ERROR(RegCreateKeyEx(
+        HKEY_CURRENT_USER,
+        subKey.c_str(),
+        0,
+        nullptr /* lpClass */,
+        REG_OPTION_NON_VOLATILE,
+        KEY_ALL_ACCESS,
+        nullptr /* lpSecurityAttributes */,
+        &hKey,
+        nullptr /* lpdwDisposition */));
+
+    bool useDefaultAssets{ false };
+
+    wil::unique_cotaskmem_string displayName;
+    wil::unique_cotaskmem_string iconFilePath;
+
+    // Try the following techniques to retrieve display name and icon:
+    // 1. From the current process.
+    // 2. Based on the best app shortcut, using the FrameworkUdk.
+    // 3. From the foreground window.
+    // 4. Use the default assets.
+    if (FAILED(RetrieveAssetsFromProcess(&displayName, &iconFilePath)) &&
+        FAILED(ToastNotifications_RetrieveAssets_Stub(&displayName, &iconFilePath)) &&
+        FAILED(RetrieveAssetsFromWindow(&displayName, &iconFilePath)))
+    {
+        displayName = wil::make_unique_string<wil::unique_cotaskmem_string>(SetDisplayNameBasedOnProcessName().c_str());
+        iconFilePath = wil::make_unique_string<wil::unique_cotaskmem_string>(defaultAppNotificationIcon);
+        useDefaultAssets = true;
+    }
+
+    // Verify if we support the provided file format for the icon. Do not verify for the default icon, we know it is supported.
+    if (!useDefaultAssets)
+    {
+        std::wstring iconFilePathAsWstring{ iconFilePath.get() };
+        auto iteratorForFileExtension{ iconFilePathAsWstring.find_first_of(L".") };
         THROW_HR_IF_MSG(E_UNEXPECTED, iteratorForFileExtension == std::wstring::npos, "You must provide a valid filepath as the app icon.");
 
-        std::wstring iconFileExtension{ iconFilePath.substr(iteratorForFileExtension) };
-        THROW_HR_IF_MSG(E_UNEXPECTED, iconFileExtension != L".ico" && iconFileExtension != L".png",
-            "You must provide a supported file extension as the icon (.ico or .png).");
+        std::wstring iconFileExtension = iconFilePathAsWstring.substr(iteratorForFileExtension);
+        THROW_HR_IF_MSG(E_UNEXPECTED,
+            iconFileExtension != L".ico" && iconFileExtension != L".png" &&
+            iconFileExtension != L".jpg" && iconFileExtension != L".bmp",
```

Comment:
These won't match ".ICO", ".JPG", ".PNG", ".BMP". Is that expected?

(usually filenames/extensions are case-insensitive in Windows)